service/src/event/service.rs

```diff
@@ -129,23 +129,72 @@ impl CeramicEventService {
         )?)
     }
 
-    pub(crate) async fn insert_events(
-        &self,
-        items: &[recon::ReconItem<EventId>],
-        source: DeliverableRequirement,
-    ) -> Result<InsertResult> {
-        let mut to_insert = Vec::new();
-        let mut invalid = Vec::new();
+    async fn validate_signed_events(
+        events: Vec<EventInsertable>,
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        // TODO: IMPLEMENT THIS
+        Ok((events, Vec::new()))
+    }
+
+    async fn validate_time_events(
+        events: Vec<EventInsertable>,
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        // TODO: IMPLEMENT THIS
+        Ok((events, Vec::new()))
+    }
+
+    pub(crate) async fn validate_events(
+        items: &[ReconItem<EventId>],
+    ) -> Result<(Vec<EventInsertable>, Vec<InvalidItem>)> {
+        let mut parsed_events = Vec::new();
+        let mut invalid_events = Vec::new();
         for event in items {
             match Self::parse_discovered_event(event).await {
-                Ok(insertable) => to_insert.push(insertable),
-                Err(err) => invalid.push(InvalidItem::InvalidFormat {
+                Ok(insertable) => parsed_events.push(insertable),
+                Err(err) => invalid_events.push(InvalidItem::InvalidFormat {
                     key: event.key.clone(),
                     reason: err.to_string(),
                 }),
             }
         }
 
+        // Group events by their type
+        let mut valid_events = Vec::new();
```

Comment:
there's a lot of allocations in this function with the way the various vectors are initialized and added to.

I'm definitely open to ideas on how to optimize this.

Reply:
I'm not too worried about the allocations yet, but we should try to use `with_capacity()` whenever we have a decent guess because it defaults to 0 and only grows by 2x, so if we have to put in 100 events we'll have to reallocate the vec with sizes 1,2,4,8,16,32,64,128.

I think we can figure out how to be smarter about allocations/resource utilization after we get things working, maybe we send on a channel in the loop, or use an arena to get nice allocations, or we restructure things a different way. Someone has to loop through and split things up, but it'd be nice if we could reuse the allocations across the different jobs (validate, order, etc).